optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -180,7 +190,332 @@
+            FilteringBiJoiner<A, A> filteringJoiner = (FilteringBiJoiner<A, A>) joiner;
+            return joinOther(otherClass)
+                    .filter(filteringJoiner.getFilter());
+        } else {
+            return joinOther(getConstraintFactory().from(otherClass), joiner);
+        }
+    }
+
+    /**
+     * As defined by {@link #joinOther(Class, BiJoiner)}.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @return a stream that matches every combination of A and a different A for which all the
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A> joiner1, BiJoiner<A, A> joiner2) {
+        return joinOther(otherClass, new BiJoiner[] {joiner1, joiner2});
+    }
+
+    /**
+     * As defined by {@link #joinOther(Class, BiJoiner)}.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @return a stream that matches every combination of A and a different A for which all the
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A> joiner1, BiJoiner<A, A> joiner2,
+            BiJoiner<A, A> joiner3) {
+        return joinOther(otherClass, new BiJoiner[] {joiner1, joiner2, joiner3});
+    }
+
+    /**
+     * As defined by {@link #joinOther(Class, BiJoiner)}.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @param joiner4 never null
+     * @return a stream that matches every combination of A and a different A for which all the
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A> joiner1, BiJoiner<A, A> joiner2,
+            BiJoiner<A, A> joiner3, BiJoiner<A, A> joiner4) {
+        return joinOther(otherClass, new BiJoiner[] {joiner1, joiner2, joiner3, joiner4});
+    }
+
+    /**
+     * As defined by {@link #joinOther(Class, BiJoiner)}.
+     * <p>
+     * This method causes <i>Unchecked generics array creation for varargs parameter</i> warnings,
+     * but we can't fix it with a {@link SafeVarargs} annotation because it's an interface method.
+     * Therefore, there are overloaded methods with up to 4 {@link BiJoiner} parameters.
+     * @param otherClass never null
+     * @param joiners never null
+     * @return a stream that matches every combination of A and a different A for which all the
+     * {@link BiJoiner joiners} are true
+     */
+    default BiConstraintStream<A, A> joinOther(Class<A> otherClass, BiJoiner<A, A>... joiners) {
+        BiJoiner<A, A> otherness = Joiners.notEqual();
+        BiJoiner[] allJoiners = Stream.concat(Stream.of(otherness), Arrays.stream(joiners))
+                .toArray(BiJoiner[]::new);
```

Comment:
The order of joiners gets changed here:
`joinOther(Foo.class, x, y z)` should be `join(Foo.class, x, y, z, notEquals())`
That's currently not the case.

Reply:
Motivation: The order of the joiners can heavily affect performance, so we must respect the user's order such that they can tweak it.

Reply:
I strongly disagree.
When using `joinOther()`, your implicit assumption is that you'll be getting only those that are not equal. Therefore, the `notEqual()` joiner *must* come first.

Reply:
(I agree on the performance point. But the fact that you'll get this joiner with `joinOther` is just assumed - if you don't want this joiner, call `join()` and do whatever you wish.)